apps/src/lib/client/tx.rs

```diff
@@ -446,36 +446,99 @@ impl masp::ShieldedUtils for CLIShieldedUtils {
 pub async fn submit_transfer(
     client: &HttpClient,
     mut ctx: Context,
-    mut args: args::TxTransfer,
+    args: args::TxTransfer,
 ) -> Result<(), tx::Error> {
-    args.tx.chain_id = args
-        .tx
-        .chain_id
-        .or_else(|| Some(ctx.config.ledger.chain_id.clone()));
-    tx::submit_transfer(client, &mut ctx.wallet, &mut ctx.shielded, args).await
+    for _ in 0..2 {
+        let arg = args.clone();
+        let (mut tx, addr, pk, tx_epoch, isf) =
+            tx::build_transfer(client, &mut ctx.wallet, &mut ctx.shielded, arg)
+            .await?;
+        // Build a transaction to reveal the signer of this transaction
+        if let Some(Address::Implicit(_)) = addr {
+            let reveal_pk = tx::build_reveal_pk(
+                client,
+                &mut ctx.wallet,
+                args::RevealPk { tx: args.tx.clone(), public_key: pk.clone() },
+            ).await?;
+            if let Some((mut rtx, _, pk)) = reveal_pk {
+                // Sign the reveal public key transaction with the fee payer
+                signing::sign_tx(&mut ctx.wallet, &mut rtx, &args.tx, &pk)
+                    .await?;
+                // Submit the reveal public key transaction first
+                tx::process_tx(client, &mut ctx.wallet, &args.tx, rtx).await?;
+                // Update the stateful PoW challenge of the outer transaction
+                #[cfg(not(feature = "mainnet"))]
+                signing::update_pow_challenge(client, &args.tx, &mut tx, &pk, isf)
+                    .await;
+            }
+        }
+        signing::sign_tx(&mut ctx.wallet, &mut tx, &args.tx, &pk)
+            .await?;
+        let result = tx::process_tx(client, &mut ctx.wallet, &args.tx, tx)
+            .await?;
+        // Query the epoch in which the transaction was probably submitted
+        let submission_epoch = rpc::query_and_print_epoch(client).await;
+
+        match result {
+            ProcessTxResponse::Applied(resp) if
+            // If a transaction is shielded
+                tx_epoch.is_some() &&
+            // And it is rejected by a VP
+                resp.code == 1.to_string() &&
+            // And the its submission epoch doesn't match construction epoch
+                tx_epoch.unwrap() != submission_epoch =>
+            {
+                // Then we probably straddled an epoch boundary. Let's retry...
+                eprintln!(
```

Comment:
we can merge this but all these `eprintln` statements make this hard to use, similar to `panics` because consumers can't catch the errors to display useful error messages to users.